[JSC] Implement Symbols as WeakMap keys

Constellation opened the pull request:
#### 290c0b91225ac430d66b4fff9a4f6606b063ba5d
<pre>
[JSC] Implement Symbols as WeakMap keys
<a href="https://bugs.webkit.org/show_bug.cgi?id=243483">https://bugs.webkit.org/show_bug.cgi?id=243483</a>

Reviewed by Ross Kirsling and Saam Barati.

This patch implements stage-3 proposal &quot;Symbols as WeakMap keys&quot;[1].
Previously, WeakMap, WeakSet, FinalizationRegistry, and WeakRef are only accepting objects as keys.
But this proposal extends it to accept non-registered Symbols too.

The runtime implementation is just changed from using JSObject* to JSCell*. And we check whether the
given value is appropriate by using canBeHeldWeakly function (specified in the spec) instead of `isObject()`.

A bit complicated part is DFG / FTL handling for WeakMap / WeakSet functions. We extend it to accept SymbolUse
and CellUse. And we emit appropriate code for that, and modifying clobberizing rules and doesGC rules too for
these new UseKind.

[1]: <a href="https://github.com/tc39/proposal-symbols-as-weakmap-keys">https://github.com/tc39/proposal-symbols-as-weakmap-keys</a>

* JSTests/stress/finalization-registry-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/v8-cleanup-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-cleanup-proxy-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-finalization-registry-basics-symbol.js: Copied from JSTests/stress/v8-finalization-registry-basics.js.
(TestConstructFinalizationRegistry):
(TestFinalizationRegistryConstructorCallAsFunction):
(TestConstructFinalizationRegistryCleanupNotCallable):
(TestConstructFinalizationRegistryWithNonCallableProxyAsCleanup):
(TestRegisterTargetAndHoldingsSameValue):
(TestRegisterWithoutFinalizationRegistry):
(TestUnregisterWithNonExistentKey):
(TestUnregisterWithNonFinalizationRegistry):
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalization-registry-basics.js:
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalizationregistry-and-weakref-symbol.js: Added.
(let.cleanup):
(setTimeout):
* JSTests/stress/v8-finalizationregistry-keeps-holdings-alive-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-finalizationregistry-scheduled-for-cleanup-multiple-times-symbol.js: Added.
(let.cleanup0):
(let.cleanup1):
(let.timeout_func):
* JSTests/stress/v8-multiple-dirty-finalization-registries-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-reentrant-gc-from-cleanup-symbol.js: Added.
(let.reentrant_gc):
(setTimeout):
* JSTests/stress/v8-stress-finalizationregistry-dirty-enqueue-symbol.js: Added.
(i.registries.push.new.FinalizationRegistry):
(registries.forEach):
* JSTests/stress/v8-undefined-holdings-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-after-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-before-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-called-twice-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup2-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup3-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-many-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-weak-unregistertoken-symbol.js: Added.
(FR.new.FinalizationRegistry):
(tryAgain):
* JSTests/stress/weak-map-constructor.js:
* JSTests/stress/weak-map-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
* JSTests/stress/weak-map-symbol.js: Added.
(shouldBe):
(test):
* JSTests/stress/weak-map-various-one-site.js: Added.
(shouldThrow):
(getTest):
(hasTest):
(setTest):
(deleteTest):
(i.test.set shouldBe):
(i.test):
* JSTests/stress/weak-map-various.js: Added.
(shouldBe):
* JSTests/stress/weak-ref-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/weak-set-constructor.js:
* JSTests/stress/weak-set-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
(test.set add):
* JSTests/stress/weak-set-symbol.js: Added.
(shouldBe):
* JSTests/stress/weak-set-various-one-site.js: Added.
(shouldThrow):
(set shouldThrow):
* JSTests/stress/weak-set-various.js: Added.
(shouldBe):
* JSTests/test262/config.yaml:
* LayoutTests/js/dom/basic-weakmap-expected.txt:
* LayoutTests/js/dom/basic-weakset-expected.txt:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::registerTarget):
(JSC::JSFinalizationRegistry::unregister):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSWeakMap.h:
* Source/JavaScriptCore/runtime/JSWeakMapInlines.h:
(JSC::JSWeakMap::set):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.cpp:
(JSC::JSWeakObjectRef::finishCreation):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.h:
* Source/JavaScriptCore/runtime/WeakMapConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapImpl.cpp:
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::takeSnapshotInternal):
(JSC::WeakMapImpl&lt;WeakMapBucket&lt;WeakMapBucketDataKey&gt;&gt;::takeSnapshot):
(JSC::WeakMapImpl&lt;WeakMapBucket&lt;WeakMapBucketDataKeyValue&gt;&gt;::takeSnapshot):
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapBucket::setKey):
(JSC::WeakMapBucket::key const):
(JSC::WeakMapBucket::deletedKey):
(JSC::WeakMapImpl::get):
(JSC::WeakMapImpl::findBucket):
(JSC::WeakMapImpl::has):
(JSC::WeakMapImpl::remove):
(JSC::WeakMapImpl::canUseBucket):
(JSC::WeakMapImpl::addInternal):
(JSC::WeakMapImpl::findBucketAlreadyHashed):
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::jsWeakMapHash):
(JSC::canBeHeldWeakly):
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::add):
* Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
(JSC::WeakMapPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapPrototype.h:
* Source/JavaScriptCore/runtime/WeakObjectRefConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
(JSC::WeakSetPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.h:

Canonical link: <a href="https://commits.webkit.org/253135@main">https://commits.webkit.org/253135@main</a>
</pre>

rkirsling commented on JSTests/stress/v8-cleanup-from-different-realm-symbol.js:
```diff
@@ -0,0 +1,36 @@
+
+// Flags: --harmony-weak-refs --expose-gc --noincremental-marking
+
+load("./resources/v8-mjsunit.js", "caller relative");
```
I guess it would have to be done in a separate patch, but what do you think about moving `stress/v8-*.js` to a `v8/` directory?

Constellation replied:
Yeah, that would be good change in a subsequent patch.

rkirsling commented on Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
```diff
@@ -31,7 +31,7 @@
 namespace JSC {
 
-const ASCIILiteral WeakMapNonObjectKeyError { "Attempted to set a non-object key in a WeakMap"_s };
+const ASCIILiteral WeakMapInvalidKeyError { "Attempted to set neither an object nor a non-registered symbol key in a WeakMap"_s };
```
This phrasing is pretty hard to read, let's make it more like:
```suggestion
const ASCIILiteral WeakMapInvalidKeyError { "WeakMap keys must be objects or non-registered symbols"_s };
```

Constellation replied:
Looks good. Changed.

rkirsling commented on Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
```diff
@@ -33,7 +33,7 @@
 namespace JSC {
 
-const ASCIILiteral WeakSetNonObjectValueError { "Attempted to add a non-object value to a WeakSet"_s };
+const ASCIILiteral WeakSetInvalidValueError { "Attempted to add neither an object nor a non-registered symbol value to a WeakSet"_s };
```
Ditto.
```suggestion
const ASCIILiteral WeakSetInvalidValueError { "WeakSet values must be objects or non-registered symbols."_s };
```

Constellation replied:
Changed.

rkirsling commented on JSTests/test262/config.yaml:
```diff
@@ -372,3 +372,12 @@ skip:
     - test/built-ins/Temporal/PlainTime/prototype/since/balance-negative-time-units.js
     - test/built-ins/Temporal/PlainTime/prototype/until/balance-negative-time-units.js
+
+    # Symbols as WeakMap keys proposal breaks existing tests https://github.com/tc39/proposal-symbols-as-weakmap-keys
```
Should we report this in a GH issue somewhere?

Constellation replied:
Maybe, we should convert it to actual testings. But for now, let's just skip them.

rkirsling commented on LayoutTests/js/dom/basic-weakmap-expected.txt:
```diff
@@ -7,13 +7,13 @@ PASS WeakMap instanceof WeakMap is false
-PASS map.set(false, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(null, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
-PASS map.set(undefined, 1) threw exception TypeError: Attempted to set a non-object key in a WeakMap.
+PASS map.set(0, 1) threw exception TypeError: Attempted to set neither an object nor a non-registered symbol key in a WeakMap.
```
How about moving these LayoutTests to JSTests/stress? :smile:

Constellation replied:
I would like to defer these tasks since we should do that for all tests in `LayoutTests/js` at the same time (and if it is tied to DOM, then we should move them to `LayoutTests/js/dom`) :)

rkirsling commented on Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp:
```diff
@@ -97,12 +98,12 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncFinalizationRegistryUnregister, (JSGlobalObjec
-    }
+    JSValue token = callFrame->argument(0);
+    if (UNLIKELY(!canBeHeldWeakly(token)))
+        return throwVMTypeError(globalObject, scope, "unregister requires an object or a non-registered symbol is the unregistration token"_s);
```
```suggestion
        return throwVMTypeError(globalObject, scope, "unregister requires an object or a non-registered symbol as the unregistration token"_s);
```

Constellation replied:
Fixed.

rkirsling commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -15290,7 +15315,13 @@ void SpeculativeJIT::compileWeakMapGet(Node* node)
     GPRReg keyGPR = key.gpr();
-    speculateObject(node->child2(), keyGPR);
+
+    // We are not checking whether Symbol is registered one or not, but it is OK since Symbol and SymbolImpl are one-on-one.
```
nit
```suggestion
    // We are not checking whether Symbol is registered one or not, but it is OK since Symbol and SymbolImpl are one-to-one.
```

Constellation replied:
Fixed.

rkirsling reviewed (approved):
r=me, though I'm not sure I fully understand the clobbering situation.

saambarati commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2564,6 +2564,11 @@ class FixupPhase : public Phase {
+            if (node->child1()->shouldSpeculateSymbol()) {
+                fixEdge<SymbolUse>(node->child1());
+                break;
+            }
```
This code already exists above :-)

Constellation replied:
Oops, removed.

saambarati commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2600,7 +2605,10 @@ class FixupPhase : public Phase {
             else
                 RELEASE_ASSERT_NOT_REACHED();
-            fixEdge<ObjectUse>(node->child2());
+            if (node->child2()->shouldSpeculateObject())
+                fixEdge<ObjectUse>(node->child2());
+            else if (node->child2()->shouldSpeculateSymbol())
+                fixEdge<SymbolUse>(node->child2());
```
since this is relying on BytecodeParser laying down CellUse, I suggest we add a validation rule for WeakMapGet where we assert we have the right kinds of speculations on edges.

Constellation replied:
Sounds good.

saambarati commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2619,14 +2627,16 @@ class FixupPhase : public Phase {
         case WeakSetAdd: {
             fixEdge<WeakSetObjectUse>(node->child1());
-            fixEdge<ObjectUse>(node->child2());
+            if (node->child2()->shouldSpeculateObject())
+                fixEdge<ObjectUse>(node->child2());
```
same comment about validation rule

Constellation replied:
Sounds good.

saambarati commented on Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
```diff
@@ -2619,14 +2627,16 @@ class FixupPhase : public Phase {
         case WeakMapSet: {
             fixEdge<WeakMapObjectUse>(m_graph.varArgChild(node, 0));
-            fixEdge<ObjectUse>(m_graph.varArgChild(node, 1));
+            if (m_graph.varArgChild(node, 1)->shouldSpeculateObject())
+                fixEdge<ObjectUse>(m_graph.varArgChild(node, 1));
```
Same comment about validation rule

Constellation replied:
Ditto.

saambarati commented on Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
```diff
@@ -76,7 +76,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapDelete, (JSGlobalObject* globalObject,
     JSValue key = callFrame->argument(0);
-    return JSValue::encode(jsBoolean(key.isObject() && map->remove(asObject(key))));
+    if (UNLIKELY(!canBeHeldWeakly(key)))
+        return JSValue::encode(jsBoolean(false));
```
Is it worth doing the extra uid load and isDerred load? Maybe it's faster to just do the hash table lookup since it's unlikely that anyone is a registered symbol. And we're guaranteed to not have the key in the set if it's registered.

saambarati replied:
Maybe we want some function for this that's different than canBeHeldWeakly since we use that necessarily along the add path

Constellation replied:
Probably, we should just check it is a cell. That's enough.

saambarati commented on Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
```diff
@@ -85,9 +87,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapGet, (JSGlobalObject* globalObject, Cal
     JSValue key = callFrame->argument(0);
-    if (!key.isObject())
+    if (UNLIKELY(!canBeHeldWeakly(key)))
         return JSValue::encode(jsUndefined());
```
ditto as above, is it even worth the isRegistered check here since it'll usually be false and just cost us extra loads

Constellation replied:
Yeah, changed it to `isCell()` check.

saambarati commented on Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
```diff
@@ -96,7 +98,9 @@ JSC_DEFINE_HOST_FUNCTION(protoFuncWeakMapHas, (JSGlobalObject* globalObject, Cal
     JSValue key = callFrame->argument(0);
-    return JSValue::encode(jsBoolean(key.isObject() && map->has(asObject(key))));
+    if (UNLIKELY(!canBeHeldWeakly(key)))
+        return JSValue::encode(jsBoolean(false));
```
ditto. Probably this happens in other access functions too, but I won't annotate them all

Constellation replied:
Changed.

saambarati commented on Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
```diff
@@ -12639,65 +12639,99 @@ IGNORE_CLANG_WARNINGS_END
+            else
+                setJSValue(vmCall(Int64, operationWeakSetGet, m_vmValue, weakMap, key, hash));
+            break;
+        }
```
Why slow path here? Why can't we run the above path?

Constellation replied:
Yeah, removed.

saambarati commented on Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
```diff
@@ -15256,6 +15256,31 @@ void SpeculativeJIT::compileMapSet(Node* node)
+        m_jit.exceptionCheck();
+        jsValueResult(resultRegs, node);
+        return;
+    }
```
ditto as comment about FTL code, why can't we run the fast path here?
If it's some bad cell value, like a string or a register symbol, we won't find it in the hash map 
Maybe just a comment to explain why it's ok

Constellation replied:
Yeah, right. Removed this.

saambarati commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -1952,7 +1952,8 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
         read(heap);
-        def(HeapLocation(WeakMapGetLoc, heap, mapEdge, keyEdge), LazyNode(node));
+        if (keyEdge.useKind() == ObjectUse || keyEdge.useKind() == SymbolUse)
+            def(HeapLocation(WeakMapGetLoc, heap, mapEdge, keyEdge), LazyNode(node));
```
Why can't we always def here, even for CellUse? If we have a string or registered symbol, its result is false

Constellation replied:
Yeah, let's def it.

saambarati commented on Source/JavaScriptCore/dfg/DFGClobberize.h:
```diff
@@ -1975,17 +1976,23 @@ void clobberize(Graph& graph, Node* node, const ReadFunctor& read, const WriteFu
+        if (keyEdge.useKind() == ObjectUse) {
+            write(JSWeakSetFields);
+            def(HeapLocation(WeakMapGetLoc, JSWeakSetFields, mapEdge, keyEdge), LazyNode(keyEdge.node()));
+        } else
```
This one makes sense to me, since we might throw if it's a registered symbol. But I wonder if we can still CSE it? Also, maybe we can add a speculation like NonRegisterdSymbol at some point.

Constellation replied:
I think probably we can just `read(World)` and `write(SideState)` too.
Then we will not reorder this node with the other one which writes something. (ensuring that exception is thrown in user observable way).
So I think we can still define heap location. Changed.

Constellation committed:
[JSC] Implement Symbols as WeakMap keys
https://bugs.webkit.org/show_bug.cgi?id=243483

Reviewed by Ross Kirsling and Saam Barati.

This patch implements stage-3 proposal "Symbols as WeakMap keys"[1].
Previously, WeakMap, WeakSet, FinalizationRegistry, and WeakRef are only accepting objects as keys.
But this proposal extends it to accept non-registered Symbols too.

The runtime implementation is just changed from using JSObject* to JSCell*. And we check whether the
given value is appropriate by using canBeHeldWeakly function (specified in the spec) instead of `isObject()`.

A bit complicated part is DFG / FTL handling for WeakMap / WeakSet functions. We extend it to accept SymbolUse
and CellUse. And we emit appropriate code for that, and modifying clobberizing rules and doesGC rules too for
these new UseKind.

[1]: https://github.com/tc39/proposal-symbols-as-weakmap-keys

* JSTests/stress/finalization-registry-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/v8-cleanup-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-cleanup-proxy-from-different-realm-symbol.js: Added.
(let.timeout_func):
* JSTests/stress/v8-finalization-registry-basics-symbol.js: Copied from JSTests/stress/v8-finalization-registry-basics.js.
(TestConstructFinalizationRegistry):
(TestFinalizationRegistryConstructorCallAsFunction):
(TestConstructFinalizationRegistryCleanupNotCallable):
(TestConstructFinalizationRegistryWithNonCallableProxyAsCleanup):
(TestRegisterTargetAndHoldingsSameValue):
(TestRegisterWithoutFinalizationRegistry):
(TestUnregisterWithNonExistentKey):
(TestUnregisterWithNonFinalizationRegistry):
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalization-registry-basics.js:
(TestWeakRefConstructorWithNonObject):
* JSTests/stress/v8-finalizationregistry-and-weakref-symbol.js: Added.
(let.cleanup):
(setTimeout):
* JSTests/stress/v8-finalizationregistry-keeps-holdings-alive-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-finalizationregistry-scheduled-for-cleanup-multiple-times-symbol.js: Added.
(let.cleanup0):
(let.cleanup1):
(let.timeout_func):
* JSTests/stress/v8-multiple-dirty-finalization-registries-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-reentrant-gc-from-cleanup-symbol.js: Added.
(let.reentrant_gc):
(setTimeout):
* JSTests/stress/v8-stress-finalizationregistry-dirty-enqueue-symbol.js: Added.
(i.registries.push.new.FinalizationRegistry):
(registries.forEach):
* JSTests/stress/v8-undefined-holdings-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-after-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-before-cleanup-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-called-twice-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup2-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-inside-cleanup3-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-unregister-many-symbol.js: Added.
(let.cleanup):
(let.timeout_func):
* JSTests/stress/v8-weak-unregistertoken-symbol.js: Added.
(FR.new.FinalizationRegistry):
(tryAgain):
* JSTests/stress/weak-map-constructor.js:
* JSTests/stress/weak-map-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
* JSTests/stress/weak-map-symbol.js: Added.
(shouldBe):
(test):
* JSTests/stress/weak-map-various-one-site.js: Added.
(shouldThrow):
(getTest):
(hasTest):
(setTest):
(deleteTest):
(i.test.set shouldBe):
(i.test):
* JSTests/stress/weak-map-various.js: Added.
(shouldBe):
* JSTests/stress/weak-ref-registered-symbol.js: Added.
(shouldThrow):
* JSTests/stress/weak-set-constructor.js:
* JSTests/stress/weak-set-registered-symbol.js: Added.
(shouldBe):
(shouldThrow):
(test.set add):
* JSTests/stress/weak-set-symbol.js: Added.
(shouldBe):
* JSTests/stress/weak-set-various-one-site.js: Added.
(shouldThrow):
(set shouldThrow):
* JSTests/stress/weak-set-various.js: Added.
(shouldBe):
* JSTests/test262/config.yaml:
* LayoutTests/js/dom/basic-weakmap-expected.txt:
* LayoutTests/js/dom/basic-weakset-expected.txt:
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/dfg/DFGOperations.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):
* Source/JavaScriptCore/runtime/FinalizationRegistryPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::registerTarget):
(JSC::JSFinalizationRegistry::unregister):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSWeakMap.h:
* Source/JavaScriptCore/runtime/JSWeakMapInlines.h:
(JSC::JSWeakMap::set):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.cpp:
(JSC::JSWeakObjectRef::finishCreation):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.h:
* Source/JavaScriptCore/runtime/WeakMapConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapImpl.cpp:
(JSC::WeakMapImpl<WeakMapBucket>::takeSnapshotInternal):
(JSC::WeakMapImpl<WeakMapBucket<WeakMapBucketDataKey>>::takeSnapshot):
(JSC::WeakMapImpl<WeakMapBucket<WeakMapBucketDataKeyValue>>::takeSnapshot):
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapBucket::setKey):
(JSC::WeakMapBucket::key const):
(JSC::WeakMapBucket::deletedKey):
(JSC::WeakMapImpl::get):
(JSC::WeakMapImpl::findBucket):
(JSC::WeakMapImpl::has):
(JSC::WeakMapImpl::remove):
(JSC::WeakMapImpl::canUseBucket):
(JSC::WeakMapImpl::addInternal):
(JSC::WeakMapImpl::findBucketAlreadyHashed):
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::jsWeakMapHash):
(JSC::canBeHeldWeakly):
(JSC::WeakMapImpl<WeakMapBucket>::add):
* Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
(JSC::WeakMapPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakMapPrototype.h:
* Source/JavaScriptCore/runtime/WeakObjectRefConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.cpp:
(JSC::WeakSetPrototype::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/WeakSetPrototype.h:

Canonical link: https://commits.webkit.org/253135@main
